File: src/java/magmac/app/compile/rule/TypeRule.java
package magmac.app.compile.rule;

import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.context.Context;
import magmac.app.compile.error.context.NodeContext;
import magmac.app.compile.error.context.StringContext;
import magmac.app.compile.error.error.CompileError;
import magmac.app.compile.error.error.CompileErrors;
import magmac.app.compile.node.Node;
import magmac.app.error.ImmutableCompileError;

public record TypeRule(String type, Rule childRule) implements Rule {
    private CompileError createError(Context context, CompileError err) {
        return new ImmutableCompileError("Cannot use type '" + this.type + "'", context, Lists.of(err));
    }

    @Override
    public CompileResult<Node> lex(String input) {
        return this.childRule.lex(input)
                .mapValue((Node node) -> node.retype(this.type))
                .mapErr((CompileError err) -> this.createError(new StringContext(input), err));
    }

    @Override
    public CompileResult<String> generate(Node node) {
        if (node.is(this.type)) {
            return this.childRule.generate(node)
                    .mapErr((CompileError err) -> this.createError(new NodeContext(node), err));
        }

        return CompileErrors.createNodeError("Type '" + this.type + "' not present", node);
    }
}

File: src/web/magmac/app/compile/rule/TypeRule.ts
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { CompileResults } from "../../../../magmac/app/compile/error/CompileResults";
import { Context } from "../../../../magmac/app/compile/error/context/Context";
import { NodeContext } from "../../../../magmac/app/compile/error/context/NodeContext";
import { StringContext } from "../../../../magmac/app/compile/error/context/StringContext";
import { CompileError } from "../../../../magmac/app/compile/error/error/CompileError";
import { CompileErrors } from "../../../../magmac/app/compile/error/error/CompileErrors";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { ImmutableCompileError } from "../../../../magmac/app/error/ImmutableCompileError";
export class TypeRule {
	createError(context : Context, err : CompileError) : CompileError {return 0;;}
	lex(input : String) : CompileResult<Node> {return 0;;}
	generate(node : Node) : CompileResult<String> {if(true){ return 0;;}return 0;;}
}
